packages/user/TokenStream/service/src/lib.rs

```diff
@@ -0,0 +1,306 @@
+                .getToken(self.token_id)
+                .precision
+                .try_into()
+                .unwrap()
+        }
+
+        pub fn add(decay_rate_per_million: u32, token_id: u32) -> Self {
+            let mut instance = Self::new(decay_rate_per_million, token_id);
+            instance.save();
+            instance
+        }
+
+        pub fn get(nft_id: u32) -> Option<Self> {
+            StreamTable::new().get_index_pk().get(&nft_id)
+        }
+
+        pub fn get_assert(nft_id: u32) -> Self {
+            check_some(Self::get(nft_id), "Stream does not exist")
+        }
+
+        pub fn check_is_owner(&self, account: AccountNumber) {
+            check(
+                psibase::services::nft::Wrapper::call()
+                    .getNft(self.nft_id)
+                    .owner
+                    == account,
+                format!("{} must be holder of NFT ID: {}", account, self.nft_id).as_str(),
+            );
+        }
+
+        fn decay_rate(&self) -> f64 {
+            std::f64::consts::LN_2 / self.half_life_seconds as f64
+        }
+
+        /// Total available to be claimed now (vested - already claimed)
+        pub fn balance_claimable(&self) -> Quantity {
+            self.total_vested()
+                .value
+                .saturating_sub(self.total_claimed.value)
+                .into()
+        }
+
+        /// Deposit `amount` into the bucket.
+        pub fn deposit(&mut self, amount: Quantity) {
+            self.claimable_at_last_deposit = self.total_vested();
+            self.total_deposited = self
+                .total_deposited
+                .value
+                .saturating_add(amount.value)
+                .into();
+            self.last_deposit_timestamp = TransactSvc::call().currentBlock().time.seconds();
+            self.save();
+        }
+
+        /// Claim everything currently claimable; returns the claimed amount.
+        pub fn claim(&mut self) -> Quantity {
+            let claimable = self
+                .total_vested()
+                .value
+                .saturating_sub(self.total_claimed.value);
+            self.total_claimed = self.total_claimed.value.saturating_add(claimable).into();
+            self.save();
+            claimable.into()
+        }
+
+        /// Total unclaimed (vested + vesting).
+        pub fn unclaimed_total(&self) -> Quantity {
+            self.total_deposited
+                .value
+                .saturating_sub(self.total_claimed.value)
+                .into()
+        }
+
+        pub fn total_vested(&self) -> Quantity {
+            if self.total_deposited.value == 0 {
+                return self.claimable_at_last_deposit;
+            }
+            let now = TransactSvc::call().currentBlock().time.seconds().seconds;
+            let delta_seconds = (now - self.last_deposit_timestamp.seconds) as f64;
+            let factor = (-self.decay_rate() * delta_seconds).exp();
```

Comment:
`exp` has unspecified precision. It's almost certainly acceptable, but we should decide how much precision we actually require and add a few tests with hard-coded constants to make sure. Integer multiples of the half-life are easy to verify.